File: ui/src/api/excelAgent.ts
import { API_BASE_URL } from './auth';
import { ApiError } from './dataSources';

export interface ProcessDataSourceResponse {
  schema_id: string;
  data_source_id: string;
  processing_status: string;
  is_ready_for_queries: boolean;
  workbook_title: string | null;
  workbook_purpose: string | null;
  domain: string | null;
  context_header_for_qa: string | null;
  table_count: number;
  relationship_count: number;
  formula_column_count: number;
  queryable_questions: string[];
  data_quality_notes: string[];
  processing_error: string | null;
  processed_at: string | null;
}

export interface ExcelSchemaResponse {
  id: string;
  data_source_id: string;
  processing_status: string;
  is_ready_for_queries: boolean;
  workbook_title: string | null;
  workbook_purpose: string | null;
  domain: string | null;
  context_header_for_qa: string | null;
  manifest: Record<string, unknown>;
  semantic_schema: Record<string, unknown>;
  enrichment: Record<string, unknown>;
  query_routing: Record<string, unknown>;
  queryable_questions: string[];
  data_quality_notes: string[];
  processing_error: string | null;
  created_at: string;
  updated_at: string;
  processed_at: string | null;
}

export interface SchemaInfoResponse {
  data_source_id: string;
  processing_status: string;
  is_ready_for_queries: boolean;
  workbook_title: string | null;
  workbook_purpose: string | null;
  domain: string | null;
  context_header_for_qa: string | null;
  table_count: number;
  relationship_count: number;
  formula_column_count: number;
  queryable_questions_count: number;
  has_data_quality_notes: boolean;
  has_enrichment: boolean;
}

export interface AgentLog {
  agent: string;
  message: string;
}

export interface QueryLogic {
  pandas: string;
  sql_like: string;
}

export interface InsightCard {
  title: string;
  message: string;
}

export interface SimulationResult {
  before_rows: Array<Record<string, unknown>>;
  after_rows: Array<Record<string, unknown>>;
  summary: string;
}

export interface NarrativeStep {
  title: string;
  body: string;
}

export interface Reasoning {
  intent: string | null;
  metric_column: string | null;
  dimension_column: string | null;
  aggregation_op: string | null;
  target_entity: string | null;
  selected_tables: string[];
  row_count: number;
  narrative: NarrativeStep[];
}

export interface AskQuestionResponse {
  success: boolean;
  answer: string;
  answer_text: string | null;
  answer_title: string | null;
  agent_logs: AgentLog[];
  query_logic: QueryLogic;
  supporting_data: Array<Record<string, unknown>>;
  insight_card: InsightCard | null;
  simulation: SimulationResult | null;
  reasoning: Reasoning | null;
  warnings: string[];
  selected_tables: string[];
  execution_time_ms: number;
  query_id: string;
  conversation_id: string | null;
}

export interface SuggestedQuestionsResponse {
  questions: string[];
  data_source_id: string;
}

export interface QueryHistoryItem {
  id: string;
  question: string;
  answer: unknown;
  code_used: string | null;
  success: boolean;
  error_message: string | null;
  execution_time_ms: number | null;
  iterations_used: number;
  created_at: string;
}

export interface QueryHistoryResponse {
  items: QueryHistoryItem[];
  total: number;
}

export interface ConversationMessage {
  id: string;
  role: string;
  content: string;
  code_used: string | null;
  execution_time_ms: number | null;
  is_error: boolean;
  error_message: string | null;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number;
  created_at: string;
}

export interface Conversation {
  id: string;
  data_source_id: string;
  title: string;
  is_active: boolean;
  total_input_tokens: number;
  total_output_tokens: number;
  total_cost_usd: number;
  created_at: string;
  updated_at: string;
  last_message_at: string | null;
  messages: ConversationMessage[];
}

export interface ConversationListItem {
  id: string;
  data_source_id: string;
  title: string;
  total_cost_usd: number;
  message_count: number;
  created_at: string;
  last_message_at: string | null;
}

export interface ConversationListResponse {
  items: ConversationListItem[];
  total: number;
}

export interface UsageSummaryResponse {
  period_days: number;
  total_input_tokens: number;
  total_output_tokens: number;
  total_cost_usd: number;
  total_calls: number;
  by_call_type: Record<string, { cost: number; count: number }>;
}

async function parseOrThrow(response: Response, fallbackMessage: string) {
  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new ApiError(response.status, error?.detail || fallbackMessage);
  }

  return response.json();
}

export async function processDataSource(
  accessToken: string,
  dataSourceId: string,
  forceReprocess = false
): Promise<ProcessDataSourceResponse> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/data-sources/${dataSourceId}/process`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify({ force_reprocess: forceReprocess }),
  });

  return parseOrThrow(response, 'Failed to process data source');
}

export async function getExcelSchema(
  accessToken: string,
  dataSourceId: string
): Promise<ExcelSchemaResponse> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/data-sources/${dataSourceId}/schema`, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
  });

  return parseOrThrow(response, 'Failed to get schema');
}

export async function getSchemaInfo(
  accessToken: string,
  dataSourceId: string
): Promise<SchemaInfoResponse> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/data-sources/${dataSourceId}/schema/info`, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
  });

  return parseOrThrow(response, 'Failed to get schema info');
}

export async function askQuestion(
  accessToken: string,
  dataSourceId: string,
  question: string,
  conversationId?: string | null
): Promise<AskQuestionResponse> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/data-sources/${dataSourceId}/ask`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify({
      question,
      conversation_id: conversationId || null,
    }),
  });

  return parseOrThrow(response, 'Failed to ask question');
}

export async function getSuggestedQuestions(
  accessToken: string,
  dataSourceId: string
): Promise<SuggestedQuestionsResponse> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/data-sources/${dataSourceId}/questions/suggested`, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
  });

  return parseOrThrow(response, 'Failed to get suggested questions');
}

export async function getQueryHistory(
  accessToken: string,
  dataSourceId: string,
  limit = 50
): Promise<QueryHistoryResponse> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/data-sources/${dataSourceId}/queries/history?limit=${limit}`, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
  });

  return parseOrThrow(response, 'Failed to get query history');
}

export async function listConversations(
  accessToken: string,
  dataSourceId?: string | null,
  skip = 0,
  limit = 50
): Promise<ConversationListResponse> {
  const query = new URLSearchParams();
  if (dataSourceId) query.set('data_source_id', dataSourceId);
  query.set('skip', String(skip));
  query.set('limit', String(limit));

  const response = await fetch(`${API_BASE_URL}/excel-agent/conversations?${query.toString()}`, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
  });

  return parseOrThrow(response, 'Failed to list conversations');
}

export async function getConversation(
  accessToken: string,
  conversationId: string
): Promise<Conversation> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/conversations/${conversationId}`, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
  });

  return parseOrThrow(response, 'Failed to get conversation');
}

export async function deleteConversation(
  accessToken: string,
  conversationId: string
): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/conversations/${conversationId}`, {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${accessToken}`,
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new ApiError(response.status, error?.detail || 'Failed to delete conversation');
  }
}

export async function getUsageSummary(
  accessToken: string,
  days = 30
): Promise<UsageSummaryResponse> {
  const response = await fetch(`${API_BASE_URL}/excel-agent/usage/summary?days=${days}`, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
  });

  return parseOrThrow(response, 'Failed to get usage summary');
}

File: core/api/v1/schemas/excel_agent.py
"""Schemas for the stable Excel-first agentic data intelligence API."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class ProcessDataSourceRequest(BaseModel):
    """Request to process or reprocess a workbook."""

    force_reprocess: bool = False


class ProcessDataSourceResponse(BaseModel):
    """High-level workbook processing status."""

    schema_id: str
    data_source_id: str
    processing_status: str
    is_ready_for_queries: bool
    workbook_title: str | None = None
    workbook_purpose: str | None = None
    domain: str | None = None
    context_header_for_qa: str | None = None
    table_count: int = 0
    relationship_count: int = 0
    formula_column_count: int = 0
    queryable_questions: list[str] = Field(default_factory=list)
    data_quality_notes: list[str] = Field(default_factory=list)
    processing_error: str | None = None
    processed_at: datetime | None = None


class ExcelSchemaResponse(BaseModel):
    """Full schema payload used by the UI workbook inspector."""

    id: str
    data_source_id: str
    processing_status: str
    is_ready_for_queries: bool
    workbook_title: str | None = None
    workbook_purpose: str | None = None
    domain: str | None = None
    context_header_for_qa: str | None = None
    manifest: dict[str, Any] = Field(default_factory=dict)
    semantic_schema: dict[str, Any] = Field(default_factory=dict)
    enrichment: dict[str, Any] = Field(default_factory=dict)
    query_routing: dict[str, Any] = Field(default_factory=dict)
    queryable_questions: list[str] = Field(default_factory=list)
    data_quality_notes: list[str] = Field(default_factory=list)
    processing_error: str | None = None
    created_at: datetime
    updated_at: datetime
    processed_at: datetime | None = None


class SchemaInfoResponse(BaseModel):
    """Compact workbook status summary."""

    data_source_id: str
    processing_status: str
    is_ready_for_queries: bool
    workbook_title: str | None = None
    workbook_purpose: str | None = None
    domain: str | None = None
    context_header_for_qa: str | None = None
    table_count: int = 0
    relationship_count: int = 0
    formula_column_count: int = 0
    queryable_questions_count: int = 0
    has_data_quality_notes: bool = False
    has_enrichment: bool = False


class ManifestSummaryResponse(BaseModel):
    """Workbook manifest summary."""

    sheet_count: int = 0
    sheet_names: list[str] = Field(default_factory=list)
    table_count: int = 0
    relationship_count: int = 0
    formula_column_count: int = 0
    tables: list[dict[str, Any]] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class EnrichmentResponse(BaseModel):
    """Optional workbook enrichment returned for inspection only."""

    semantic_model: dict[str, Any] = Field(default_factory=dict)
    data_quality: dict[str, Any] = Field(default_factory=dict)
    kpi_recommendations: list[dict[str, Any]] = Field(default_factory=list)
    table_overview: list[dict[str, Any]] = Field(default_factory=list)
    demo_scenarios: list[str] = Field(default_factory=list)
    relationship_graph: dict[str, Any] = Field(default_factory=dict)
    formula_catalog: dict[str, Any] = Field(default_factory=dict)
    metadata_index: dict[str, Any] = Field(default_factory=dict)


class AskQuestionRequest(BaseModel):
    """Natural-language workbook question."""

    question: str
    conversation_id: str | None = None


class AgentLogResponse(BaseModel):
    """One visible agent action in the orchestrator pipeline."""

    agent: str
    message: str


class QueryLogicResponse(BaseModel):
    """Transparent execution representation."""

    pandas: str = ""
    sql_like: str = ""


class InsightCardResponse(BaseModel):
    """Short explainable insight surfaced by the UI."""

    title: str
    message: str


class SimulationResponse(BaseModel):
    """Scenario before/after comparison."""

    before_rows: list[dict[str, Any]] = Field(default_factory=list)
    after_rows: list[dict[str, Any]] = Field(default_factory=list)
    summary: str


class NarrativeStep(BaseModel):
    """A single plain-English step in the reasoning narrative."""

    title: str
    body: str


class ReasoningResponse(BaseModel):
    """Structured reasoning payload explaining how the answer was derived."""

    intent: str | None = None
    metric_column: str | None = None
    dimension_column: str | None = None
    aggregation_op: str | None = None
    target_entity: str | None = None
    selected_tables: list[str] = Field(default_factory=list)
    row_count: int = 0
    narrative: list[NarrativeStep] = Field(default_factory=list)


class AskQuestionResponse(BaseModel):
    """Stable orchestrator response for the demo UI."""

    success: bool
    answer: str
    answer_text: str | None = None
    answer_title: str | None = None
    agent_logs: list[AgentLogResponse] = Field(default_factory=list)
    query_logic: QueryLogicResponse = Field(default_factory=QueryLogicResponse)
    supporting_data: list[dict[str, Any]] = Field(default_factory=list)
    insight_card: InsightCardResponse | None = None
    simulation: SimulationResponse | None = None
    reasoning: ReasoningResponse | None = None
    warnings: list[str] = Field(default_factory=list)
    selected_tables: list[str] = Field(default_factory=list)
    execution_time_ms: int = 0
    query_id: str
    conversation_id: str | None = None


class SuggestedQuestionsResponse(BaseModel):
    """Suggested workbook questions."""

    questions: list[str] = Field(default_factory=list)
    data_source_id: str


class QueryHistoryItem(BaseModel):
    """Stored query history item."""

    id: str
    question: str
    answer: Any = None
    code_used: str | None = None
    success: bool
    error_message: str | None = None
    execution_time_ms: int | None = None
    iterations_used: int = 0
    created_at: datetime


class QueryHistoryResponse(BaseModel):
    """Query history list."""

    items: list[QueryHistoryItem] = Field(default_factory=list)
    total: int


class ConversationMessageResponse(BaseModel):
    """Conversation message entry."""

    id: str
    role: str
    content: str
    code_used: str | None = None
    execution_time_ms: int | None = None
    is_error: bool
    error_message: str | None = None
    input_tokens: int = 0
    output_tokens: int = 0
    cost_usd: float = 0.0
    created_at: datetime


class ConversationResponse(BaseModel):
    """Full conversation payload."""

    id: str
    data_source_id: str
    title: str
    is_active: bool
    total_input_tokens: int
    total_output_tokens: int
    total_cost_usd: float
    created_at: datetime
    updated_at: datetime
    last_message_at: datetime | None = None
    messages: list[ConversationMessageResponse] = Field(default_factory=list)


class ConversationListItem(BaseModel):
    """Conversation row for list views."""

    id: str
    data_source_id: str
    title: str
    total_cost_usd: float
    message_count: int
    created_at: datetime
    last_message_at: datetime | None = None


class ConversationListResponse(BaseModel):
    """Conversation listing payload."""

    items: list[ConversationListItem] = Field(default_factory=list)
    total: int


class UsageSummaryResponse(BaseModel):
    """Usage summary payload."""

    period_days: int
    total_input_tokens: int
    total_output_tokens: int
    total_cost_usd: float
    total_calls: int
    by_call_type: dict[str, dict[str, float | int]] = Field(default_factory=dict)
